add qwen 3 chat template deep dive (#2834)
huggingface/blog: qwen-3-chat-template-deep-dive.md

    {{- '<think>\n\n</think>\n\n' }}
{%- endif %}
```

QwQ for example, forces reasoning in every conversation.

```jinja
{# QwQ #}
{%- if add_generation_prompt %}
    {{- '<|im_start|>assistant\n<think>\n' }}
{%- endif %}
```

If the `enable_thinking` is true, the model is able to decide whether to think or not. 

You can test test out the template with the following code:

```js
import { Template } from "@huggingface/jinja";
import { downloadFile } from "@huggingface/hub";

const HF_TOKEN = process.env.HF_TOKEN;

const file = await downloadFile({
  repo: "Qwen/Qwen3-235B-A22B",
  path: "tokenizer_config.json",
  accessToken: HF_TOKEN,
});
const config = await file!.json();

const template = new Template(config.chat_template);
const result = template.render({
  messages,
  add_generation_prompt: true,
  enable_thinking: false,  
  bos_token: config.bos_token,
  eos_token: config.eos_token,
});
```

## 2. Context Management Should be Dynamic

_Qwen-3 utilizes a rolling checkpoint system, intelligently preserving or pruning reasoning blocks to maintain relevant context. Older models discarded reasoning prematurely to save tokens._ 

> Qwen-3 introduces a "**_rolling checkpoint_**" by traversing the message list in reverse to find the latest user turn that wasn’t a tool echo. For any assistant replies after that index it keeps the full `<think>` blocks; everything earlier is stripped out.

Reviewer: nit:
```suggestion
Qwen-3 introduces a "**_rolling checkpoint_**" by traversing the message list in reverse to find the latest user turn that wasn’t a tool call. For any assistant replies after that index it keeps the full `<think>` blocks; everything earlier is stripped out.
```

**Why this matters**:
- Keeps the active plan visible during a multi‑step tool call.
- Supports nested tool workflows without losing context.
- Saves tokens by pruning thoughts the model no longer needs.
- Prevents "stale" reasoning from bleeding into new tasks.

### Example

Here's an example of chain-of-thought preservation through tool calls with Qwen-3 and QwQ.
![image/png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/qwen-3-chat-template/qwen-chat-output.png)
<p style="text-align:center; font-style:italic; font-size:medium;">
  Check out <a href="https://www.npmjs.com/package/@huggingface/jinja">@huggingface/jinja</a> for testing out the chat templates
</p>

## 3. Tool Arguments Need Better Serialization

Before, every `tool_call.arguments` field was piped through ` | tojson`, even if it was already a JSON‑encoded string—risking double‑escaping. Qwen‑3 checks the type first and only serializes when necessary.

```jinja
{# Qwen3 #}
{%- if tool_call.arguments is string %}
    {{- tool_call.arguments }}
{%- else %}
    {{- tool_call.arguments | tojson }}
{%- endif %}
```

## 4. Default System Prompts Are Unnecessary

> Qwen‑2.5 automatically inserted a default Alibaba system prompt:

Author: Yeah maybe I can word it differently. Do you think there's better way to position this point?

> Qwen‑2.5 automatically inserted a default Alibaba system prompt:

Author: reworded it a bit

**Why this matters**:
- Keeps the active plan visible during a multi‑step tool call.
- Supports nested tool workflows without losing context.
- Saves tokens by pruning thoughts the model no longer needs.
- Prevents "stale" reasoning from bleeding into new tasks.

### Example

Here's an example of chain-of-thought preservation through tool calls with Qwen-3 and QwQ.
![image/png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/qwen-3-chat-template/qwen-chat-output.png)
<p style="text-align:center; font-style:italic; font-size:medium;">
  Check out <a href="https://www.npmjs.com/package/@huggingface/jinja">@huggingface/jinja</a> for testing out the chat templates
</p>

## 3. Tool Arguments Need Better Serialization

Before, every `tool_call.arguments` field was piped through ` | tojson`, even if it was already a JSON‑encoded string—risking double‑escaping. Qwen‑3 checks the type first and only serializes when necessary.

```jinja
{# Qwen3 #}
{%- if tool_call.arguments is string %}
    {{- tool_call.arguments }}
{%- else %}
    {{- tool_call.arguments | tojson }}
{%- endif %}
```

## 4. No Need for a Default System Prompt

> Like many models, the Qwen‑2.5 series had a default Alibaba system prompt:

Author: ```suggestion
Like many models, the Qwen‑2.5 series has a default system prompt.
```

> You are Qwen, created by Alibaba Cloud. You are a helpful assistant.

> This is pretty common to help models to respond to user questions like "Who are you?".

Author: ```suggestion
This is pretty common as it helps models respond to user questions like "Who are you?".
```

> Qwen-3 (alongside QwQ) ships without a default system prompt, but still answers the "Who are you?" question correctly without it.

Author: ```suggestion
Qwen-3 (and QwQ) ships without a default system prompt. Despite this, the model can still accurately identify its creator. 
```

thumbnail: /blog/assets/qwen-3-chat-template-deep-dive/thumbnail.jpg
authors:
- user: cfahlgren1
---

# The 4 Things Qwen-3’s Chat Template Teaches Us

_**What a boring Jinja snippet tells us about the new Qwen-3 model.**_

The new Qwen-3 model by [Qwen](https://huggingface.co/qwen) ships with a much more sophisticated chat template than it's predecessors Qwen-2.5 and QwQ. By taking a look at the differences in the Jinja template, we can find interesting insights into the new model.

<h2 style="text-align: center; margin-bottom: 0.5rem; font-style: italic;">Chat Templates</h2>
<ul style="display: flex; justify-content: center; list-style: none; padding: 0; margin: 0;">
  <li style="margin-right: 1rem;"><a href="https://huggingface.co/Qwen/Qwen3-235B-A22B?chat_template=default">Qwen-3 Chat Template</a></li>
  <li style="margin-right: 1rem;"><a href="https://huggingface.co/Qwen/Qwen2.5-Coder-32B-Instruct?chat_template=default">Qwen-2.5 Chat Template</a></li>
  <li><a href="https://huggingface.co/Qwen/QwQ-32B?chat_template=default">Qwen-QwQ Chat Template</a></li>
</ul>

## 1. Reasoning doesn't have to be forced

_**and you can make it optional via a simple prefill...**_

> Qwen-3 is unique in it's ability to toggle reasoning via the `enable_thinking` flag. When set to false, the template inserts an empty <think></think> pair, telling the model to skip step‑by‑step thoughts. Earlier models baked the <think> tag into every generation, forcing chain‑of‑thought whether you wanted it or not.

Reviewer 2: You may need to escape these tags?
`&lt;think&gt;&lt;/think&gt;`

Author: good catch
